### ISSUE TITLE
clear `src` attribute on recycled `<img>` nodes

### ISSUE THREAD

**Issue by robbiewxyz** [developit/preact#351]
Here's [a codepen](http://codepen.io/robbiewxyz/pen/xEAzLB) with the problem:  when the `src` attribute of an image is changed, the old picture will keep showing while the new picture is loading.  This behavior persists even if the image is removed from the DOM.  When combined with Preact's node-caching, this can get crazy, with old pictures showing up again in unrelated places.

Clearing the image with `.removeAttribute('src')` should solve the issue.  I'll plan on putting together a pull request here shorty.

**Comment by developit** [comment]
Seems like there are two identical issues. Want to delete one of them? 😛

**Comment by robbiewxyz** [comment]
Oh, oops.  I was on a bad internet connection.  Can I actually delete an issue or just close it?

**Comment by mikestead** [comment]
@developit I'm seeing something similar with the latest Preact (8.1.0). If I comment out the contents of Preact's `collectComponent` function it goes away.

The scenario we have is a list of products, you click one and go to the detail page with an image header (`picture` element), click back, then click a new product. When you hit the product page the second time it shows the old product image for a split second. If we switch back to React all is ok again.

I've been trying for the past hour to create a smaller codepen example but no luck recreating it, and ran out of time for now I'm afraid.

**Comment by developit** [comment]
Preact still recycles components even after we removed DOM recycling. #664 proposes removing this, and I have [a branch](https://github.com/developit/preact/compare/scratch/perf) that tries to regain the optimization lost by removing the component recycler.

That said, if you're seeing the image for a really short time, that might be related to the async change in Preact 8, which has been [reverted to the 7.x behavior](https://github.com/developit/preact/commit/576dda310a8e31e524fa2ecc803d6d406597c41e) and will be in the next release.

**Comment by mikestead** [comment]
Thanks for the details, and all the hard work on Preact.

I tried reverting to pre 8.x versions and the issue still seems to be present. I then tried resetting the src element on img elements in `collectComponent` and this seemed to mostly work so I assume it's the recycling issue. I'll subscribe to #664 and keep an eye on progress 👍

**Comment by jackmoore** [comment]
I've also been running into this problem with old images showing while new images load, which is not a trivial amount of time. The old images may be visible for seconds.

I tried using `.removeAttribute('src')` on `componentWillUnmount`, which did work to start each image on a clean slate but there is some sort of race condition there because sometimes the new image would end up without an src attribute. It was unworkable.

I settled on creating this component as a replacement for `<img>` elements, which seemed to provide both a clean slate and resolve the `.removeAttribute('src')` race condition:

```js
export default class extends React.Component {
	state = {src: null};

	componentDidMount() {
		this.setState({src: this.props.src});
	}

	componentWillReceiveProps(nextProps) {
		if (nextProps.src !== this.props.src) {
			this.setState({src: nextProps.src});
		}
	}

	render() {
		return <img {...this.props} src={this.state.src}/>;
	}
}
```
